Stratis.Bitcoin.Features.RPC/Controllers/RPCController.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+using System.Linq;
+using System.Net;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Logging;
+using NBitcoin.RPC;
+using Stratis.Bitcoin.Builder.Feature;
+using Stratis.Bitcoin.Utilities.JsonErrors;
+
+namespace Stratis.Bitcoin.Features.RPC.Controllers
+{
+    /// <summary>
+    /// Controller providing API operations on the RPC feature.
+    /// </summary>
+    [Route("api/[controller]")]
+    public class RPCController : Controller
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Full Node.</summary>
+        private readonly IFullNode fullNode;
+
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="loggerFactory">Factory to be used to create logger for the node.</param>
+        public RPCController(IFullNode fullNode, ILoggerFactory loggerFactory)
+        {
+            this.fullNode = fullNode;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Finds an RPC method by name by looking in all features assemblies of the full node.
+        /// </summary>
+        /// <param name="name">The name of the RPC method to find.</param>
+        /// <returns>The 'MethodInfo' associated with the RPC method or null if the method is not found.</returns>
+        private MethodInfo GetRPCMethod(string name)
```

Comment:
I think this method is quite taxing, especially when called on a controller action.
It seems that the result of this method is only used to make sure there is such a method and that the controller exists. Only then do you actually make an RPC call.

How about you don't worry about whether the method exists and just proceeds with the call?
If an exception is called because the method doesn't exists, you handle the exception there.

Reply:
It is also used to determine the order in which to pass the query string arguments to the method (as well as providing default values if required).

Reply:
What happens if you don't pass the items in the right order in an RPC call?
In any case the responsibility is on the caller.
So if we have an api call that looks like `/api/rpc/callbyname?methodName=getblock&params=param1,param2,param3`, that should be clear enough.

Reply:
If this is to avoid the overhead of extracting the MethodInfo then I suggest it may be better to cache the method info (as I am now doing) and have the benefit of passing the arguments by name as per our earlier discussion:

`1) Add a "callbyname" HttpGet method to an RPCController to enable calls such as: http://localhost:37220/api/rpc/callbyname?methodname=getblock&height=1112`

Reply:
Is that all it takes, 6ms? That's fast.
Here is another suggestion (might be outside of the scope of this change but let's think about it):
what if we could 
1. cache all the RPC methods when the controller gets initialized, instead of caching on demand
2. then provide another api call that lists all the available/discovered RPC commands
For that we'd probably have to add an attribute to the RPC methods to say that they are RPC in order to avoid having non-RPC methods referenced.

What do you think?